crates/task-impls/src/quorum_vote/handlers.rs

```diff
@@ -135,6 +135,52 @@ async fn handle_quorum_proposal_validated_drb_calculation_seed<
     Ok(())
 }
 
+/// Handles calling add_epoch_root and sync_l1 on Membership if necessary.
+async fn handle_quorum_proposal_validated_add_epoch_root<
+    TYPES: NodeType,
+    I: NodeImplementation<TYPES>,
+    V: Versions,
+>(
+    proposal: &QuorumProposal2<TYPES>,
+    task_state: &mut QuorumVoteTaskState<TYPES, I, V>,
+    leaf_views: &[LeafInfo<TYPES>],
+) {
+    let decided_block_number = leaf_views
+        .last()
+        .unwrap()
+        .leaf
+        .block_header()
+        .block_number();
+
+    // Skip if this is not the expected block.
+    if task_state.epoch_height != 0 && (decided_block_number + 3) % task_state.epoch_height == 0 {
```

Comment:
Currently what would happen if a node misses the block when the update is triggered? Would it never have `add_epoch_root` called for that epoch? 

I'm wondering if we the Membership trait should have a method like `is_epoch_initialized(&self, epoch) -> bool` so that if the node is not online for the beginning of the epoch we can still initialize the membership for the epoch.

Or can we handle this well enough with catchup only in the confirmation layer? What do you think @imabdulbasit ?

Reply:
I think doing a catchup is better, but I am wondering how we can do that now with Hotshot managing the locks

Reply:
For our existing catchup it's hotshot that triggers the catchup (for example by calling `validate_and_apply_header`) and the confirmation layer that performs the catchup. Maybe it should be the same here: hotshot needs to call something to trigger catchup for the membership and the sequencer performs the catchup. 

If we were to do it analoguous for membership we would have to make all the reading functions async and &mut self which seems terrible. I think we can think of something better.

Reply:
For example hotshot could call `add_epoch_root` (probably with another name) for every view then the sequencer could decide if it needs to do catchup for this epoch (and in most cases do (almost) nothing and return `None` as write_callback). This way we ensure the catchup is actually triggered.

I think our overall design is a bit sub-optimal because it's an invariant that the membership is static for an epoch but that invariant isn't really reflected in the code. I think eventually we should enforce this in hotshot and the confirmation layer should somehow provide a constructor for a "EpochMembership" type that takes an `epoch` as input that is called by hotshot where needed. @ss-es mentioned this would be difficult to do in hotshot but I'm not sure pushing the complexity elsewhere is actually better.

Reply:
yeah but then hotshot has to decide when to call it? I think it would calling it for every view would be too much maybe. It makes sense to do catch up when the membership get functions() do not return any data?

Reply:
maybe we do catchup at the node startup and pass the Committee so hotshot does not need to do anything?

Reply:
I think catchup at startup might not work if a node misses some views for some other reason, for example due to a temporary network outage.

Reply:
> yeah but then hotshot has to decide when to call it? I think it would calling it for every view would be too much maybe. 

If it doesn't do anything except for once per epoch then what's the problem with calling it every view?

Reply:
> I'm wondering if  the Membership trait should have a method like `is_epoch_initialized(&self, epoch) -> bool` so that if the node is not online for the beginning of the epoch we can still initialize the membership for the epoch. But then we also have to dig out the correct header from history which I think might be cumbersome.

Why do we need this? When node comes online it should go through same routine as all nodes, which includes calling `add_epoch_root` for the block being proposed/validated.

Reply:
I might be reading it wrongly but I think the current implementation in this PR only calls the function if `decided_block_number + 3) % task_state.epoch_height == 0` so if it joins at a later point and the decided block has moved on it won't call it. Maybe I'm misunderstanding.